File: front/src/api/group.ts
import { withApiBase } from './base';

export interface GroupMember {
    id: number;
    group_id: number;
    member_id: string;
    member_type: 'user' | 'friend';
    name?: string;
    avatar?: string;
    join_time: string;
}

export interface GroupRead {
    id: number;
    name: string;
    avatar?: string;
    description?: string;
    owner_id: string;
    auto_reply: boolean;
    member_count?: number;
    members?: GroupMember[];
    create_time: string;
    update_time: string;
}

export interface GroupReadWithMembers extends GroupRead {
    members: GroupMember[];
}

export interface GroupCreate {
    name: string;
    member_ids: string[]; // List of persona_ids
    description?: string;
    avatar?: string;
    auto_reply?: boolean;
}

export interface GroupUpdate {
    name?: string;
    description?: string;
    avatar?: string;
    auto_reply?: boolean;
}

export interface GroupMessageRead {
    id: number;
    group_id: number;
    sender_id: string;
    sender_type: 'user' | 'friend';
    content: string;
    message_type: 'text' | 'system' | '@';
    mentions?: string[];
    create_time: string;
}

async function request<T>(options: { url: string; method: string; data?: any; params?: any }): Promise<T> {
    const url = new URL(withApiBase(`/api${options.url}`));
    if (options.params) {
        Object.keys(options.params).forEach((key) => url.searchParams.append(key, options.params[key]));
    }

    const response = await fetch(url.toString(), {
        method: options.method,
        headers: {
            'Content-Type': 'application/json',
        },
        body: options.data ? JSON.stringify(options.data) : undefined,
    });

    if (!response.ok) {
        const errorBody = await response.json().catch(() => ({}));
        throw new Error(errorBody.detail || `Request failed with status ${response.status}`);
    }

    return response.json();
}

export const groupApi = {
    /**
     * 获取用户加入的所有群组
     */
    getGroups: () => {
        return request<GroupRead[]>({
            url: '/groups',
            method: 'GET',
        });
    },

    /**
     * 创建群组
     */
    createGroup: (data: GroupCreate) => {
        return request<GroupRead>({
            url: '/group/create',
            method: 'POST',
            data,
        });
    },

    /**
     * 获取群详情及成员列表
     */
    getGroup: (id: number) => {
        return request<GroupReadWithMembers>({
            url: `/group/${id}`,
            method: 'GET',
        });
    },

    /**
     * 邀请新成员
     */
    inviteMembers: (groupId: number, memberIds: string[]) => {
        return request<boolean>({
            url: '/group/invite',
            method: 'POST',
            data: {
                group_id: groupId,
                member_ids: memberIds,
            },
        });
    },

    /**
     * 更新群设置
     */
    updateSettings: (groupId: number, settings: GroupUpdate) => {
        return request<GroupRead>({
            url: '/group/settings',
            method: 'PATCH',
            data: {
                group_id: groupId,
                group_in: settings,
            },
        });
    },

    /**
     * 退出群组
     */
    exitGroup: (groupId: number) => {
        return request<boolean>({
            url: '/group/exit',
            method: 'DELETE',
            params: { group_id: groupId },
        });
    },

    /**
     * 移除群成员
     */
    removeMember: (groupId: number, memberId: string) => {
        return request<boolean>({
            url: `/group/${groupId}/member/${memberId}`,
            method: 'DELETE',
        });
    },

    /**
     * 获取群组历史消息
     */
    getGroupMessages: (groupId: number, skip: number = 0, limit: number = 100) => {
        return request<GroupMessageRead[]>({
            url: `/group/${groupId}/messages`,
            method: 'GET',
            params: { skip, limit },
        });
    },

    /**
     * 清空群消息记录
     */
    clearMessages: (groupId: number) => {
        return request<{ message: string }>({
            url: `/chat/group/${groupId}/messages`,
            method: 'DELETE',
        });
    },
    /**
     * 发送群消息并流式获取响应
     */
    async *sendGroupMessageStream(groupId: number, data: { content: string, mentions?: string[], enable_thinking?: boolean }): AsyncGenerator<{ event: string, data: any }> {
        const url = withApiBase(`/api/chat/group/${groupId}/messages`);
        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(data),
        });

        if (!response.ok) {
            const errorBody = await response.json().catch(() => ({}));
            throw new Error(errorBody.detail || `Request failed with status ${response.status}`);
        }

        const reader = response.body?.getReader();
        if (!reader) return;

        const decoder = new TextDecoder();
        let buffer = '';

        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done) break;

                buffer += decoder.decode(value, { stream: true });
                const parts = buffer.split('\n\n');
                buffer = parts.pop() || '';

                for (const part of parts) {
                    const lines = part.split('\n');
                    let eventType = 'message';
                    let dataString = '';

                    for (const line of lines) {
                        if (line.startsWith('event: ')) {
                            eventType = line.slice(7).trim();
                        } else if (line.startsWith('data: ')) {
                            dataString += (dataString ? '\n' : '') + line.slice(6);
                        }
                    }

                    if (dataString) {
                        try {
                            yield {
                                event: eventType,
                                data: JSON.parse(dataString),
                            };
                        } catch (e) {
                            console.error('Failed to parse SSE data', e);
                            yield { event: eventType, data: dataString };
                        }
                    }
                }
            }
        } finally {
            reader.releaseLock();
        }
    },
};

File: server/app/api/endpoints/group.py
from fastapi import APIRouter, Depends, HTTPException, Query, Body
from sqlalchemy.orm import Session
from typing import List

from app.api import deps
from app.schemas import group as group_schemas
from app.services.group_service import group_service
from app.services.memo.constants import DEFAULT_USER_ID

router = APIRouter()

@router.get("/groups", response_model=List[group_schemas.GroupReadWithMembers])
def read_user_groups(
    db: Session = Depends(deps.get_db),
):
    """
    获取当前用户加入的所有群组。
    """
    return group_service.get_user_groups(db)

@router.post("/group/create", response_model=group_schemas.GroupReadWithMembers)
def create_group(
    *,
    db: Session = Depends(deps.get_db),
    group_in: group_schemas.GroupCreate,
):
    """
    创建新群组。
    """
    return group_service.create_group(db, group_in)

@router.get("/group/{id}", response_model=group_schemas.GroupReadWithMembers)
def read_group(
    *,
    db: Session = Depends(deps.get_db),
    id: int,
):
    """
    获取群详情及成员列表。
    """
    group = group_service.get_group(db, id)
    if not group:
        raise HTTPException(status_code=404, detail="Group not found")
    
    # Check if user is a member
    if not group_service.is_member(db, id, DEFAULT_USER_ID):
        raise HTTPException(status_code=403, detail="Not a member of this group")
        
    return group

@router.post("/group/invite", response_model=bool)
def invite_members(
    *,
    db: Session = Depends(deps.get_db),
    group_id: int = Body(...),
    member_ids: List[str] = Body(...),
):
    """
    向现有群组添加新成员。
    """
    success = group_service.add_members(db, group_id, member_ids)
    if not success:
        raise HTTPException(status_code=403, detail="Failed to invite members")
    return True

@router.patch("/group/settings", response_model=group_schemas.GroupRead)
def update_group_settings(
    *,
    db: Session = Depends(deps.get_db),
    group_id: int = Body(...),
    group_in: group_schemas.GroupUpdate = Body(...),
):
    """
    更新群设置。仅群主可操作。
    """
    group = group_service.update_group(db, group_id, group_in)
    if not group:
        raise HTTPException(status_code=403, detail="Not authorized or group not found")
    return group

@router.delete("/group/exit", response_model=bool)
def exit_group(
    *,
    db: Session = Depends(deps.get_db),
    group_id: int = Query(...),
):
    """
    当前用户退出群组。
    """
    success = group_service.exit_group(db, group_id)
    if not success:
        raise HTTPException(status_code=400, detail="Failed to exit group")
    return True

@router.delete("/group/{id}/member/{member_id}", response_model=bool)
def remove_group_member(
    *,
    db: Session = Depends(deps.get_db),
    id: int,
    member_id: str,
):
    """
    群主移除指定群成员。
    """
    success = group_service.remove_member(db, id, member_id)
    if not success:
        raise HTTPException(status_code=403, detail="Not authorized or member not found")
    return True

# --- Group Chat Messaging ---

@router.get("/group/{id}/messages", response_model=List[group_schemas.GroupMessageRead])
def read_group_messages(
    *,
    db: Session = Depends(deps.get_db),
    id: int,
    skip: int = 0,
    limit: int = 100,
):
    """
    获取群组历史消息。
    """
    from app.models.group import GroupMessage, GroupMember
    
    # 鉴权：检查当前用户是否在群组中
    member = db.query(GroupMember).filter(
        GroupMember.group_id == id,
        GroupMember.member_id == DEFAULT_USER_ID,
        GroupMember.member_type == "user"
    ).first()
    if not member:
        raise HTTPException(status_code=403, detail="Not a member of this group")

    messages = db.query(GroupMessage).filter(GroupMessage.group_id == id).order_by(GroupMessage.create_time.desc()).offset(skip).limit(limit).all()
    return list(reversed(messages))


